---
title: "oopsRBPs\n 01: Quality Control and Aggregation to Proteins"
author: "Tom Smith, Veronica Dezi, Manasa Ramakrishna"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  pdf_document:
    toc: yes
    toc_depth: '3'
  html_document:
    toc: yes
    toc_depth: 3
    toc_float: yes
---

```{r global_options, echo=FALSE}
knitr::opts_chunk$set(fig.width=12,fig.height=8,warning=FALSE, message=FALSE,tidy = TRUE,tidy.opts=list(width.cutoff=50))
```
## 1. Introduction  
This is a large experiment by Veronica testing the effects of Insulin deprivation and stimulation on MCF10A "normal" breast cells that are highly dependent on insulin for their growth. The experimental set up is shown below. As you can see, there are 5 replicates for each conditions. Set1 and Set 2 have been analysed in two 10-plex TMT experiments on the mass spectrometer. Here, we would like to assess the differences in RNA-binding proteins, if any, in (1) Insulin-starved and (2) Insulin-stimulated cells relative to their controls.  

In the following lines of code, we will:  
 1. Look the raw data  
 2. Filter data to remove those mapping to multiple proteins 
 3. Aggregate the data into protein-level quantification  
 4. Filter the data to remove contaminants  
 5. Normalise the data (if needed)  
 6. Perform QC checks 
 7. Look for differentially expressed RBPs across conditions
 8. Look for GO term enrichment in the differential RBPs

![Experimental setup for MCF10A cells stimulated by/starved of Insulin](../MCF10A-Insulin-Experimental-Setup.png)
***
We start by installing and loading the libraries required for our analysis. Make sure you have downloaded the GitHub repository https://github.com/TomSmithCGAT/CamProt_R locally. Also make sure you are in the directory that contains the Rmd notebooks prior to running this code, otherwise program will not run. Raw data is in the folder "../raw", plots will be printed out to the folder "../plots" and results in the form of tables will be printed out to the folder "../results". Large annotation files necessary for GO term enrichment are in the folder "../shared_files". 

```{r 00_Startup}

# Libraries needed for analysis
suppressMessages(library(MSnbase))
suppressMessages(library(dplyr))

# source file that contains useful functions for analysis
source("../../CamProt_R/Utility.R", chdir=TRUE)

# Source file that contains a list of glycoproteins that have been identified in previous OOPS experiments. I'm not convinced about removing these proteins from the study at the ourset.
glycoproteins <- read.table("../shared_files/glycoproteins.tsv", sep="\t", header=TRUE)
glycoproteins <- glycoproteins %>% pull(protein) %>% as.character() %>% unique()
```
A set of functions needed within this script.
```{r 01_Functions}

#--------------------------------------------------------
# Function  : run_parse_features
# Aim       : Function-to-parse-features-for-all-data
#--------------------------------------------------------

run_parse_features <- function(fname){
  
  df = parse_features(fname,master_protein_col='Master.Protein.Accessions',protein_col='Master.Protein.Accessions')
  df_noglyc = df[!df$Master.Protein.Accessions %in% glycoproteins,]
  
  cat(sprintf("%s\tExcluding glycoproteins\n", length(rownames(df_noglyc))))
  cat(sprintf("\nThese peptides are associated with %s master proteins\n", 
              length(unique(df_noglyc$Master.Protein.Accessions))))
  
  return(list(df = df, df_noglyc = df_noglyc))
}

#----------------------------------------------------------------------------------
# Function  : assess_rbp_tot_overlap
# Aim       : quantify overlap between total protein and rbp (oops) experiments
#----------------------------------------------------------------------------------
assess_rbp_tot_overlap <- function(rbp_df, tot_df, suf){
  print(paste("RBP master proteins also in total",suf,"=",length(intersect(unique(tot_df$Master.Protein.Accessions),
                               unique(rbp_df$Master.Protein.Accessions))),sep=" "))
  print(paste("RBP master proteins NOT in total",suf,"=",
              length(setdiff(unique(rbp_df$Master.Protein.Accessions),
                             unique(tot_df$Master.Protein.Accessions)))))
  
  print(paste("Percentage RBP master proteins in total",suf,"=",round(length(intersect(unique(tot_df$Master.Protein.Accessions),unique(rbp_df$Master.Protein.Accessions)))*100/length(unique(rbp_df$Master.Protein.Accessions)),2),"%"))
}

#----------------------------------------------------------------------------------
# Function  : summariseMissing
# Aim       : print tally of missing values for each study across all peptides/prots
#----------------------------------------------------------------------------------
summariseMissing <- function(res){
  cat("\ntallies for missing data (# samples with missing)")
  print(table(rowSums(is.na(as.data.frame(exprs(res))))))
}

```
## 2. Read in the peptide-level quantification  
What we obtain following a mass spectrometer run is peptide-level quantification from Proteome Discoverer(PD). Here we perform some Quality Control checks before aggregating data to protein-level quantification. 

As discussed before, there are 2 sets of experiments that need to be analysed :  
a. Non-Insulin Stimulated (4hr-Starved) vs Insulin Stimulated (30min.Insulin) MCF10A cells
b. Unstarved vs Starved (4hr-Starved) MCF10A cells  

The function 'parse_features' reads in a peptide-level data file and processes it remove any contaminants, peptides missing a protein assignment and those that are missing quantification values (details in 'Utility.R'). 

```{r 02_Peptide-level-quantification}
# Parsing input files
files = grep("reannot",list.files("../raw",full.names = T),value=T)
suff = gsub("_reannot.txt|_peptides","",sapply(strsplit(files,"/"),"[[",3))

# Process all experimental data
for(f in 1:length(files)){
  print(toupper(suff[f])) 
  r = run_parse_features(files[f])
  assign(paste(suff[f],"df",sep="_"),r$df)
  assign(paste(suff[f],"df_noglyc",sep="_"),r$df_noglyc)
}

# List objects that have been created
grep("df",ls(),value=T)

# Summarise the number of RBPs and Total proteins in each set of experiments

# 1. Non-insulin (4hr-Starved) vs Insulin Stimulated (30min-Insulin)
assess_rbp_tot_overlap(rbp_ni_df, total_ni_df, "N-vs-I")

# 2. Unstarved vs Starved (4hr-Starved)
assess_rbp_tot_overlap(rbp_us_df, total_us_df, "U-vs-S")
```
In the Non-Stimulated vs Insulin Stimulated experiments, nearly 80% of RBPs are also in the total samples. In the Unstarved vs Starved experiments, only about 70% of RBPs are also in the total samples. 

For most of the RBPs we will also have information regarding the changes in the total abundance for the protein, regardless of the abundance of the RNA-bound fraction. However, for 20-30% of them, we won't. 

Let's start aggregating the quantification.First we create the dataframes to make an MSnSet object...

```{r 03_Creating-an-MSnSet-from-data}

# Read in experimental set-up data
samples_ni_inf <- "../raw/samples_ni.tsv"
samples_us_inf <- "../raw/samples_us.tsv"

total_ni_res <- makeMSNSet(total_ni_df, samples_ni_inf)
rbp_ni_res <- makeMSNSet(rbp_ni_df, samples_ni_inf)

total_us_res <- makeMSNSet(total_us_df, samples_us_inf)
rbp_us_res <- makeMSNSet(rbp_us_df, samples_us_inf)

# Check relative quantificationd data
p <- plotLabelQuant(total_ni_res, log=T)
p <- plotLabelQuant(rbp_ni_res, log=T)
p <- plotLabelQuant(total_us_res, log=T)
p <- plotLabelQuant(rbp_us_res, log=T)
```
For the Total N_I experiment, 4hr-Starved_1 seems to be sticking out. For the Total & RBP U_S experiment, 4hr-Starved_4 seems to be sticking out. 

Aggregate over the peptides with the same sequence but different modifications. Don't worry about the errors. Missing values are correctly handled here.
```{r}

total_ni_res_pep_agg <- agg_to_peptides(total_ni_res)
rbp_ni_res_pep_agg <- agg_to_peptides(rbp_ni_res)

total_us_res_pep_agg <- agg_to_peptides(total_us_res)
rbp_us_res_pep_agg <- agg_to_peptides(rbp_us_res)

saveRDS(total_ni_res_pep_agg, file="../results/total_ni_res_pep_agg.rds")
saveRDS(rbp_ni_res_pep_agg, file="../results/rbp_ni_res_pep_agg.rds")
saveRDS(total_us_res_pep_agg, file="../results/total_us_res_pep_agg.rds")
saveRDS(rbp_us_res_pep_agg, file="../results/rbp_us_res_pep_agg.rds")

```

Re-check missing values
```{r}
summariseMissing(total_ni_res_pep_agg)
summariseMissing(rbp_ni_res_pep_agg)
summariseMissing(total_us_res_pep_agg)
summariseMissing(rbp_us_res_pep_agg)
```
OK, so very few missing values now. This is what we would expect. Now, we aggregate from peptide-level to protein level abundance

```{r}
total_ni_res_pro_agg <- agg_to_protein(total_ni_res_pep_agg, "Master.Protein.Accessions")
rbp_ni_res_pro_agg <- agg_to_protein(rbp_ni_res_pep_agg, "Master.Protein.Accessions")

total_us_res_pro_agg <- agg_to_protein(total_us_res_pep_agg, "Master.Protein.Accessions")
rbp_us_res_pro_agg <- agg_to_protein(rbp_us_res_pep_agg, "Master.Protein.Accessions")

saveRDS(total_ni_res_pro_agg, file="../results/total_ni_res_pro_agg.rds")
saveRDS(rbp_ni_res_pro_agg, file="../results/rbp_ni_res_pro_agg.rds")
saveRDS(total_us_res_pro_agg, file="../results/total_us_res_pro_agg.rds")
saveRDS(rbp_us_res_pro_agg, file="../results/rbp_us_res_pro_agg.rds")
```

And summarise the missing values again. 
```{r}
summariseMissing(total_ni_res_pro_agg)
summariseMissing(rbp_ni_res_pro_agg)
summariseMissing(total_us_res_pro_agg)
summariseMissing(rbp_us_res_pro_agg)
```
Note, we're now down to very few missing values. So few that I'd recommend we simply remove these proteins entirely (0/2792 Total NI proteins; 1/1399 RBP NI proteins; 3/2805 for Total US proteins, 33/1048). Since the maximal missing values is 3%, we will remove proteins with missing values for now. Only exception might be RBP U_S in case we are loosing important ones. 
```{r}
total_ni_res_pro_agg <- total_ni_res_pro_agg[rowSums(is.na(exprs(total_ni_res_pro_agg)))==0,]
rbp_ni_res_pro_agg <- rbp_ni_res_pro_agg[rowSums(is.na(exprs(rbp_ni_res_pro_agg)))==0,]
total_us_res_pro_agg <- total_us_res_pro_agg[rowSums(is.na(exprs(total_us_res_pro_agg)))==0,]
rbp_us_res_pro_agg <- rbp_us_res_pro_agg[rowSums(is.na(exprs(rbp_us_res_pro_agg)))==0,]

summariseMissing(total_ni_res_pro_agg)
summariseMissing(rbp_ni_res_pro_agg)
summariseMissing(total_us_res_pro_agg)
summariseMissing(rbp_us_res_pro_agg)
```

Below, we check how many peptides we have per protein.
```{r}

plotPepCounts <- function(res_pep_agg){
fData(res_pep_agg) <- fData(res_pep_agg)[,-grep("CV.*", colnames(fData(res_pep_agg)))]

gb <- fData(res_pep_agg)$Master.Protein.Accessions
gb <- factor(gb, levels = unique(gb))
peptide_counts <- as.data.frame(table(gb))

print(sprintf("Out of %s master proteins, we have %s one-hit wonder proteins (%s %%)",
              length(unique(gb)), sum(peptide_counts==1),
              round(100*sum(peptide_counts==1)/length(unique(gb)), 1)))

p <- ggplot(peptide_counts, aes(log(Freq,2))) + geom_histogram() + my_theme
return(p)
}

p1=plotPepCounts(total_ni_res_pep_agg)
p2=plotPepCounts(rbp_ni_res_pep_agg)
p3=plotPepCounts(total_us_res_pep_agg)
p4=plotPepCounts(rbp_us_res_pep_agg)

grid.arrange(p1,p2,p3,p4,nrow=2)
```
So we have 13-19% "one-hit" wonders across the 4 experiments (NI - Total,RBP; US - Total,RBP). We'll leave these in for now but we could consider excluding these as we have less confident abundance estimates here

Next, we log center normalise the data. The total amount of peptide labelled in each sample should be the same so we can assume any difference in the total intensity per tag is due to inaccuracies in peptide quantification prior to labelling, pipetting etc, or non-identical labelling efficiencies. Whatever the cause, we can normalise the tags since we want to obtain a relative measure between tags. The plots below show before and after normalisation
```{r}
total_ni_res_pro_agg_norm <- logCenterNormalise(total_ni_res_pro_agg)
exprs(total_ni_res_pro_agg_norm) <- exprs(total_ni_res_pro_agg_norm) + median(log2(exprs(total_ni_res_pro_agg)))
  
rbp_ni_res_pro_agg_norm <- logCenterNormalise(rbp_ni_res_pro_agg)
exprs(rbp_ni_res_pro_agg_norm) <- exprs(rbp_ni_res_pro_agg_norm) + median(log2(exprs(rbp_ni_res_pro_agg)))

total_us_res_pro_agg_norm <- logCenterNormalise(total_us_res_pro_agg)
exprs(total_us_res_pro_agg_norm) <- exprs(total_us_res_pro_agg_norm) + median(log2(exprs(total_us_res_pro_agg)))
  
rbp_us_res_pro_agg_norm <- logCenterNormalise(rbp_us_res_pro_agg)
exprs(rbp_us_res_pro_agg_norm) <- exprs(rbp_us_res_pro_agg_norm) + median(log2(exprs(rbp_us_res_pro_agg)))
```


```{r}
pa <- plotLabelQuant(total_ni_res_pro_agg, log=TRUE)
pb <- plotLabelQuant(rbp_ni_res_pro_agg, log=TRUE)

pc <- plotLabelQuant(total_ni_res_pro_agg_norm, log=FALSE)
pd <- plotLabelQuant(rbp_ni_res_pro_agg_norm, log=FALSE)

pe <- plotLabelQuant(total_us_res_pro_agg, log=TRUE)
pf <- plotLabelQuant(rbp_us_res_pro_agg, log=TRUE)

pg <- plotLabelQuant(total_us_res_pro_agg_norm, log=FALSE)
ph <- plotLabelQuant(rbp_us_res_pro_agg_norm, log=FALSE)

grid.arrange(pa$p2,pb$p2,pc$p2,pd$p2,pe$p2,pf$p2,pg$p2,ph$p2,nrow=4)
```
```

```{r}
saveRDS(total_ni_res_pro_agg_norm, file="../results/total_ni_res_pro_agg_norm")
saveRDS(rbp_ni_res_pro_agg_norm, file="../results/rbp_ni_res_pro_agg_norm")
saveRDS(total_us_res_pro_agg_norm, file="../results/total_us_res_pro_agg_norm")
saveRDS(rbp_us_res_pro_agg_norm, file="../results/rbp_us_res_pro_agg_norm")
```


