---
title: "03: Changes in RNA Binding"
author: "Eneko Villanueva,Rayner Queiroz, Manasa Ramakrishna, Tom Smith"
date: '`r format(Sys.Date(), "%B %d, %Y")`'
output:
  pdf_document:
    toc: yes
    toc_depth: '3'
  html_document:
    toc: yes
    toc_depth: 3
    toc_float: yes
editor_options: 
  chunk_output_type: inline
---

```{r global_options, echo=FALSE}
knitr::opts_chunk$set(fig.width=12,fig.height=8,warASng=FALSE, message=FALSE,tidy = TRUE,tidy.opts=list(width.cutoff=50))
```
## 1. Introduction  

In this section of the code, we are finally doing the interesting analysis which is finding out if there are any RBPs that are differentially expressed between conditions. Having looked at the data thus far, the extreme variability of the RBP Unstarved vs Starved samples might mean that we cannot really do a differential analysis with that set. However, we'll give it a go and see what happen. 

```{r 00_Reading-in-utility-function, eval=T, echo=F, warning=FALSE}

source("../../CamProt_R/Utility.R")

library(plyr)
library(dplyr)
library(tidyverse)
library(limma)
library(biobroom)
library(Hmisc)
library(MSnbase)

# set up standardised plotting scheme
theme_set(theme_bw(base_size = 20) +
            theme(panel.grid.major=element_blank(),
                  panel.grid.minor=element_blank(),
                  aspect.ratio=1))

cbPalette <- c("#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7", "#999999")
```

```{r 01_Functions-for-analysis, eval=T, echo=F}

# A set of functions needed within this script.

#-----------------------------------------------------------------
# Function  : make_design_matrix
# Aim       : make a design matrix from a sample data file
# Input     : data frame that contains a column called "Condition"
#------------------------------------------------------------------

# Design matrix for various comparisons
make_design_matrix<-function(setup){
  des = model.matrix(~setup$Condition)
  rownames(des) = rownames(setup)
  colnames(des) = gsub("setup\\$Condition","",colnames(des))
  print(des)
}

#-----------------------------------------------------------------
# Function  : my_volcano_plot
# Aim       : make a volcano plot from topTable object 
# Input     : data frame that contains a column called "Condition"
#------------------------------------------------------------------
my_volcanoplot <- function(topTableResults){
  p <- topTableResults %>%
    mutate(sig=ifelse(adj.P.Val<0.05, "sig.", "not sig.")) %>% # add "sig" column
    ggplot(aes(logFC, -log10(P.Value), colour=sig)) +
    geom_point(size=0.25) +
    scale_colour_manual(values=c("black", cbPalette[6]), name="") # manually adjust colours
  
  return(p)
}

#-----------------------------------------------------------------
# Function  : combine_esets
# Aim       : combine Total and RBP datasets for limma analysis 
# Input     : Two MSnSets - one for Total Proteome and one for OOPS
# Output    : One combined MSnSet
#------------------------------------------------------------------
combine_esets <- function(total,rbp){

  # make the column names for the two MSnSets uASque
  colnames(total) <- paste0(colnames(total), "_Total")
  colnames(rbp) <- paste0(colnames(rbp), "_OOPS")
  
  # make the ExpressionSet
  combined_intensities <- ExpressionSet(cbind(exprs(total), exprs(rbp)))
  
  # Add the feature data
  fData(combined_intensities) <- fData(total)
  
  # Add the phenotype data
  pData(combined_intensities) <- rbind(pData(total), pData(rbp))
  pData(combined_intensities)$Condition <- factor(sapply(strsplit(pData(combined_intensities)$Sample_name,"_"),"[[",1))
  pData(combined_intensities)$Type <- factor(sapply(strsplit(rownames(pData(combined_intensities)),"_"),"[[",2),level=c("Total","OOPS"))
  
  return(combined_intensities)
}


#------------------------------------------------------------------------------
# Function  : modify_output
# Aim       : modify output from TopTable to just keep informative columns
# Input     : output from 'run_limma'
# Output    : simplified data frame with results of linear modelling
#-------------------------------------------------------------------------------


modify_output <- function(rbps_de,cols=c("Master.Protein.Accessions",
       "Master.Protein.Descriptions",
       "logFC", "CI.L", "CI.R", "AveExpr",
       "t", "P.Value","adj.P.Val","B")){
  rbps_mod = rbps_de[,cols]
  rbps_mod$protein_desc = sapply(strsplit(rbps_mod$Master.Protein.Descriptions,"OS="),"[[",1)
  rbps_mod$gene_name = sapply(strsplit(rbps_mod$Master.Protein.Descriptions,"GN=|PE="),"[[",2)
  colnames(rbps_mod)[1] = "uniprot_id"
  return(rbps_mod)
}

#------------------------------------------------------------------------------
# Function  : run_limma
# Aim       : Runs limma on combined OOPS and Total protein data
# Input     : Combined dataset from section 4b and coefficient of interest
# Output    : limma results
#-------------------------------------------------------------------------------

run_limma <- function(combined_intensities,coeff){
  
  condition <- combined_intensities$Condition
  type <- combined_intensities$Type
  sample_name <- combined_intensities$Sample_name
  
  # limma design matrix
  design <- model.matrix(~0+condition*type)
  
  # Fit a linear model 
  rna_binding_fit <- lmFit(combined_intensities, design)
  
  # Draw contrasts
  rna_binding_fit <- contrasts.fit(rna_binding_fit, coefficients=coeff)
  
  # Compute moderated statistics and log odds of DE using empirical bayes moderation
  rna_binding_fit <- eBayes(rna_binding_fit, trend=TRUE, robust=TRUE)
  
  # Look at the data
  plotSA(rna_binding_fit)
  
  # Plotting sigASficance
  rna_binding_p_value_status <- ifelse(rna_binding_fit$p.value[,coeff]<0.05, "sig", "not_sig")
  limma::plotMA(rna_binding_fit, status=rna_binding_p_value_status, values=c("sig", "not_sig"),
                col=c("magenta", "black"), cex=c(0.8,0.2), main="")
  
  # Count DE proteins
  summary(decideTests(rna_binding_fit, p.value=0.05, adjust.method="BH"))
  
  # Volcano plot of all results
  all_rna_binding_results <- topTable(rna_binding_fit, coef = coeff, n = Inf, confint=TRUE)
  my_volcanoplot(all_rna_binding_results)
  
  # Return all results
  return(all_rna_binding_results)

}

#------------------------------------------------------------------------------
# Function  : simple_limma
# Aim       : Runs limma on a single dataset - oops or total protein
# Input     : MSnSet, design matrix and coefficient of interest
# Output    : limma results
#-------------------------------------------------------------------------------

simple_limma <- function(protein_quant,des,coeff){
  
  # fit linear model to each protein
  total_fit_lm <- lmFit(exprs(protein_quant), des)
  
  # extract results for coefficient of interest
  total_fit_lm_c <- contrasts.fit(total_fit_lm, coefficients=coeff)
  
  # Relationship between mean and variance
  # As expected, there is a relationship between mean intensity and variance, although this is almost all limited to the very low intensity values having high variance. 
  #plotSA(total_fit_lm_c)
  
  # Below, we perform the emperical Bayesian shrinking of the std errors towards the trend (`trend=TRUE`). We will also use the `robust=TRUE` option to ensure that the outliers don't affect the trend estimation.

  # shrink std errors to abundance vs. stdev trend
  total_fit_lm_e_c <- eBayes(total_fit_lm_c, trend=TRUE, robust=TRUE)
  #plotSA(total_fit_lm_e_c)
  
  # Identify significant changes
  # Below, we plot the average intensity vs log change. This is a useful QC plot to show that nothing odd has happened with our linear modeling.
  p_value_status <- ifelse(total_fit_lm_e_c$p.value[,coeff]<0.05, "sig", "not_sig")
  
  # plot
  limma::plotMA(total_fit_lm_e_c, status=p_value_status,col=c("magenta", "black"), cex=c(0.8,0.1), main="")
  
  # Count DE proteins
  summary(decideTests(total_fit_lm_e_c, p.value=0.05, adjust.method="BH"))
  
  # Note that most of these changes are relatively slight (<2-fold)
  # Extract all results from limma (n=Inf)
  all_results <- topTable(total_fit_lm_e_c, coef = coeff, n = Inf)
  
  # we'll make a coupld of volcano plots so easier to wrap this up into a function
  my_volcanoplot(all_results)
  
  return(all_results)
}

#------------------------------------------------------------------------------
# Function  : makeLongExprs
# Aim       : Convert an expression matrix into the long form for plotting
# Input     : MSnSet, list of proteins
# Output    : Long form of expression data with variables in columns 
#-------------------------------------------------------------------------------
makeLongExprs <- function(obj, intersecting_proteins){
  
  obj_exprs <- exprs(obj[intersecting_proteins,])
  colnames(obj_exprs) <- pData(obj)$Sample_name
  
  long_exprs <- obj_exprs %>%
    data.frame() %>%
    tibble::rownames_to_column("uniprotID") %>%
    gather(key="Sample", value="Intensity", -uniprotID) %>%
    separate(Sample, into=c("Condition", "Replicate"),sep="_")
  
  long_exprs$Condition = factor(long_exprs$Condition,levels=c("Ctrl","X100uM.Arsenite","X400uM.Arsenite"))
  
  return(long_exprs)
}

#------------------------------------------------------------------------------
# Function  : plotTop10
# Aim       : Plots Total vs RBP trend plots for a set of proteins
# Input     : 
#       combined_exprs : combined total and rbp expression in long form, 
#       lowest_p_proteins : list of proteins, ideally with smallest p-value
#       suf : Suffix used in the title of plots explaiASng which experiment it is
# Output    : Long form of expression data with variables in columns 
#-------------------------------------------------------------------------------
plotTop10<-function(combined_exprs,lowest_p_proteins,suf){
  test_df<- combined_exprs %>% 
            filter(uniprotID %in% lowest_p_proteins) %>%
            arrange(uniprotID, Condition, Type)
  test_df$Type = factor(test_df$Type,levels = c("RBPS","Total"))
  test_df$Entry.name <- gsub("_HUMAN", "", test_df$Entry.name)
  
  test_df %>%
    ggplot(aes(Condition, Intensity)) +
    geom_point() +
    stat_summary(fun.y="mean", geom="line", aes(group=Type)) +
    facet_grid(Entry.name~Type,scales = "free_y") +
    xlab("") +
    ylab("Protein abundance (log2)") +
    theme(axis.text.x=element_text(size=12, angle=90, vjust=0.5, hjust=1),
          legend.position="bottom",
          title = element_text(vjust = 1,hjust=0.5,size=12))+labs(title=suf)
}

```
## 2. Reading in normalised, outlier-free data

We start by reading in the normalised data and then setting up for a Limma analysis

```{r 02_Reading-in-data}

total_as_protein_quant <- readRDS("../results/total_as_res_pro_agg_norm")
oops_as_protein_quant <- readRDS("../results/rbp_as_res_pro_agg_norm")
```

## 3. LIMMA for differential protein expression analysis
LIMMA stands for Linear Models for Microarray and RNA-Seq Data and is a package used for the analysis of gene expression data from microarrays or RNAseq experiments. It's major selling point is that it is able to use linear models to assess differential expression in the context of multifactor designed experiments. Rather usefully, limma does distinguish data to be "from proteins" or "from RNA" which makes it quite handy to apply to Proteomics data.There are a few steps to DE analysis by limma.  
1. Create a data matrix with samples in columns and proteins in rows. We can use the "exprs" slot in an MSnSet for this. 
2. Create a design matrix that tells limma about samples, conditions and replicates. We can use the `pData` from MSnSet for this.  
3. Fit a linear model to the data(1) using the design(2).  
4. Define contrasts of interest i.e which gruops of samples you want to test for differential protein expression.    
5. Extract results for the contrast of interest.  
6. Look at the top proteins.  

Initially, we perform this analysis for each of the 4 datasets separately.

## 3a. Combining total and RBP data

It is relatively easy to perform a pairwise comparison between treated and untreated samples either in the RBP or Total proteome. What about changes in RNA binding? For this, we need combine the two MSnSets into a single ExpressionSet. We start by intersecting proteins within the Arsenite experiments so we can compare just those proteins that are captures across both total and RBP datasets. 

```{r 3a_Intersecting-proteins}
intersecting_as_proteins <- intersect(rownames(total_as_protein_quant), rownames(oops_as_protein_quant))
print(paste("Number of RBPs also captured in the Total Proteome for Control vs Arsenite treated samples is", length(intersecting_as_proteins)),sep="")

# Subset of intersecting AS proteins only
total_as_for_combination <- total_as_protein_quant[intersecting_as_proteins,]
rbp_as_for_combination <- oops_as_protein_quant[intersecting_as_proteins,]
```

```{r 3b_Combining-expression-data}

combined_as_intensities = combine_esets(total_as_for_combination,rbp_as_for_combination)
pData(combined_as_intensities)$Condition = factor(pData(combined_as_intensities)$Condition,levels = c("Ctrl","100uM-Arsenite","400uM-Arsenite"))

```

The we run `limma` on the combined intensities and this time test for a signficant interaction coefficient. There are 224 proteins differentially expressed in cells treated with 100uM NaAs2 relative to Control and similarly, there are 184 proteins differentially expressed in cells treated with 400uM NaAs2 relative to Control.

```{r}
coefficients=c("Condition100uM.Arsenite.TypeOOPS",
                                          "Condition400uM.Arsenite.TypeOOPS")

  fit$t <- NULL
  fit$p.value <- NULL
  fit$lods <- NULL
  fit$F <- NULL
  fit$F.p.value <- NULL
  ncoef <- NCOL(fit$coefficients)
  if (!is.null(contrasts)) {
      contrasts <- as.matrix(contrasts)
      rn <- rownames(contrasts)
      cn <- colnames(fit$coefficients)
      if (!is.null(rn) && !is.null(cn) && any(rn != cn)) 
          warning("row names of contrasts don't match col names of coefficients")
  }
  else {
      ncont <- length(coefficients)
      contrasts <- diag(ncoef)
      rownames(contrasts) <- colnames(contrasts) <- colnames(fit$coefficients)
      contrasts <- contrasts[, coefficients, drop = FALSE]
  }
```

```{r 3c_Limma-on-combined-data}


# Create a design matrix
#treat <- paste(combined_as_intensities$Condition, combined_as_intensities$Type, sep=".") 
#design = model.matrix(~0+treat)
##treat = gsub("\\-",".",treat)
condition <- combined_as_intensities$Condition
type <- combined_as_intensities$Type
design<-model.matrix(~0+condition*type)
colnames(design) <- gsub("[:-]", ".", colnames(design))
# Calculate correlation between replicates
corfit <- duplicateCorrelation(combined_as_intensities,design2,block=combined_as_intensities$Replicate)

# Fit a linear model to include a block and correlation
fit <- lmFit(combined_as_intensities,design,correlation=corfit$consensus)

# Create a vector to extract contrasts
#cm <- makeContrasts(
#  Ctrlvs100uM = (treat100uM.Arsenite.Total-treatCtrl.Total)-(treat100uM.Arsenite.OOPS-treatCtrl.OOPS),
#  Ctrlvs400uM = (treat400uM.Arsenite.Total-treat400uM.Arsenite.OOPS)-(treatCtrl.Total-treatCtrl.OOPS),
#  As400vs100uM = (treat100uM.Arsenite.Total-treat100uM.Arsenite.OOPS)-(treat400uM.Arsenite.Total-treat400uM.Arsenite.OOPS),
#  levels=design)



# Extract contrasts from the fitted linear model
fit2 <- contrasts.fit(fit, coefficients=c("condition100uM.Arsenite.typeOOPS",
                                          "condition400uM.Arsenite.typeOOPS"))

# Calculating the F-statistic
fit2 <- eBayes(fit2)

# Extracting the top hits for each contrast
Ctrl.100uM = topTable(fit2, adjust = "BH",coef = "condition100uM.Arsenite.typeOOPS",number = Inf,confint = T)
de_mod_100uM = modify_output(Ctrl.100uM)
p_value_100uM <- de_mod_100uM %>% filter(P.Value <= 0.05)
write.table(p_value_100uM, "../results/Ctrl-vs-100uM-Arsenite-Treated-BLOCK-rawp-le-0.05.tsv",sep="\t",row.names=F,quote=F)

Ctrl.400uM = topTable(fit2, adjust = "BH",coef = "condition400uM.Arsenite.typeOOPS",number = Inf,confint = T)
de_mod_400uM = modify_output(Ctrl.400uM)
p_value_400uM <- de_mod_400uM %>% filter(P.Value <= 0.05)
write.table(p_value_400uM,"../results/Ctrl-vs-400uM-Arsenite-Treated-rawp-le-0.05.tsv",sep="\t",row.names=F,quote=F)

#As400.100uM = topTable(fit2, adjust = "BH",coef = "As400vs100uM",number = Inf,confint = T)
#de_mod_100.400uM = modify_output(As400.100uM)
#p_value_100.400uM <- de_mod_100.400uM %>% filter(P.Value <= 0.05) 
#write.table(p_value_100.400uM,"../results/100-vs-400uM-Arsenite-Treated-rawp-le-0.05.tsv",sep="\t",row.names=F,quote=F)

# Significantly DE proteins - only a few
de_mod_100uM %>% filter(adj.P.Val <= 0.05)
de_mod_400uM %>% filter(adj.P.Val <= 0.05)
#de_mod_100.400uM %>% filter(adj.P.Val <= 0.05)

saveRDS(Ctrl.100uM, "../results/Ctrl.100uM.rds")
saveRDS(Ctrl.400uM, "../results/Ctrl.400uM.rds")
#saveRDS(As400.100uM, "../results/As400.100uM.rds")
```
```{r}
Ctrl.400uM["P10515",] %>% modify_output()
plotTop10(combined_as_exprs,c("P10515"),"")
```

We are also interested in working out if there is any difference between the 100uM and 400uM treated cells. I have set the design up slightly differently here incorporating the replicate number as a blocking factor. Blocking can be applied in any situation where there are batch effects or where the experiment has been conducted in blocks. The treatments can be adjusted for differences between the blocks by using a model formula of the form  

**_design <- model.matrix(~Block+Treatment)_**  


```{r 3d_Alternate-design}

# Create a design matrix
treat <- paste(combined_as_intensities$Condition, combined_as_intensities$Type, sep=".") 
treat = gsub("\\-",".",treat)
design2 = model.matrix(~0+treat)

# Calculate correlation between replicates
corfit <- duplicateCorrelation(combined_as_intensities,design2,block=combined_as_intensities$Rep)

# Fit a linear model to include a block and correlation
fit <- lmFit(combined_as_intensities,design2,block=pData(combined_as_intensities)$Replicate,correlation=corfit$consensus)

# Create a vector to extract contrasts
cm <- makeContrasts(
  Ctrlvs100uM = (treatCtrl.Total-treatCtrl.OOPS)-(treat100uM.Arsenite.Total-treat100uM.Arsenite.OOPS),
  Ctrlvs400uM = (treatCtrl.Total-treatCtrl.OOPS)-(treat400uM.Arsenite.Total-treat400uM.Arsenite.OOPS),
  As400vs100uM = (treat400uM.Arsenite.Total-treat400uM.Arsenite.OOPS)-(treat100uM.Arsenite.Total-treat100uM.Arsenite.OOPS),
  levels=design2)

# Extract contrasts from the fitted linear model
fit2 <- contrasts.fit(fit, cm)

# Calculating the F-statistic
fit2 <- eBayes(fit2)

# Extracting the top hits for each contrast
Ctrl.100uM = topTable(fit2, adjust = "BH",coef = "Ctrlvs100uM",number = Inf,confint = T)
de_mod_100uM = modify_output(Ctrl.100uM)
p_value_100uM <- de_mod_100uM %>% filter(P.Value <= 0.05)
write.table(p_value_100uM, "../results/Ctrl-vs-100uM-Arsenite-Treated-BLOCK-rawp-le-0.05.tsv",sep="\t",row.names=F,quote=F)

Ctrl.400uM = topTable(fit2, adjust = "BH",coef = "Ctrlvs400uM",number = Inf,confint = T)
de_mod_400uM = modify_output(Ctrl.400uM)
p_value_400uM <- de_mod_400uM %>% filter(P.Value <= 0.05)
write.table(p_value_400uM,"../results/Ctrl-vs-400uM-Arsenite-Treated-BLOCK-rawp-le-0.05.tsv",sep="\t",row.names=F,quote=F)

As400.100uM = topTable(fit2, adjust = "BH",coef = "As400vs100uM",number = Inf,confint = T)
de_mod_100.400uM = modify_output(As400.100uM)
p_value_100.400uM <- de_mod_100.400uM %>% filter(P.Value <= 0.05) 
write.table(p_value_100.400uM,"../results/100-vs-400uM-Arsenite-Treated-BLOCK-rawp-le-0.05.tsv",sep="\t",row.names=F,quote=F)

# Significantly DE proteins - only a few
de_mod_100uM %>% filter(adj.P.Val <= 0.05)
de_mod_400uM %>% filter(adj.P.Val <= 0.05)
de_mod_100.400uM %>% filter(adj.P.Val <= 0.05)
```




While there are several proteins whose p-value is significant, only a few survive multiple testing correction (shown above). Of these, there doesn't seem to be much in the way of a functional theme. Hence, there isn't much to go on for GO enrichment analysis. We could use the lists based just on raw p-value and see if we have any luck with them. 

## 4. Plotting expression of Top 10 proteins across studies
This is to see whether we have a reason for not spotting any DE proteins. Looking at the plots, it is pretty clear that the differences between treatments is pretty minimal i.e the trend lines are not very extreme and only in a few cases are the trend lines really varied between Total proteome and RBP-ome.  

```{r 4_Plotting-expression-data-per-protein, fig.width=12, fig.height=12}

# Combining Total and RBP data : AS
total_as_exprs <- makeLongExprs(total_as_protein_quant,intersecting_as_proteins)
oops_as_exprs <- makeLongExprs(oops_as_protein_quant,intersecting_as_proteins)
combined_as_exprs <- rbind(cbind(total_as_exprs,Type="Total"), cbind(oops_as_exprs,Type="RBPS"))

# Adding protein information
protein_info <- read.delim("../shared_files/human_protein_ids_plus_gene_names.tsv")
combined_as_exprs <- combined_as_exprs %>% merge(protein_info, by.x="uniprotID", by.y="Entry")

# Renaming the levels for better plotting
library(plyr)
combined_as_exprs$Condition = revalue(combined_as_exprs$Condition, c("X400uM.Arsenite"="As-400uM", "X100uM.Arsenite"="As-100uM"))

# Top 10 proteins from both comparisons
lowest_p_ctrl_100uM_proteins <- p_value_100uM %>% arrange(P.Value) %>% pull(uniprot_id) %>% head(10)
lowest_p_ctrl_400uM_proteins <- p_value_400uM %>% arrange(P.Value) %>% pull(uniprot_id) %>% head(10)
lowest_p_100_400uM_proteins <- p_value_100.400uM %>% arrange(P.Value) %>% pull(uniprot_id) %>% head(10)

# Plots
plotTop10(combined_as_exprs,lowest_p_ctrl_100uM_proteins,"Ctrl vs 100uM Arsenite")

plotTop10(combined_as_exprs,lowest_p_ctrl_400uM_proteins,"Ctrl vs 400uM Arsenite")


plotTop10(combined_as_exprs,lowest_p_100_400uM_proteins,"100uM vs 400uM Arsenite")
```

```{r}
plotTop10(combined_as_exprs,c("Q7KZF4", "Q8NC51"),"")




```

